sklearn/utils/_array_api.py

```diff
@@ -440,9 +439,27 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
         return float(numpy.average(sample_score_np, weights=sample_weight_np))
 
     if not xp.isdtype(sample_score.dtype, "real floating"):
-        # We move to cpu device ahead of time since certain devices may not support
-        # float64, but we want the same precision for all devices and namespaces.
-        sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
+        # We move to cpu device since certain devices may not support float64,
+        # but we want the same precision for all devices and namespaces.
+        try:
+            sample_score = xp.astype(sample_score, xp.float64)
+        except Exception:
+            if xp.__name__ not in {
+                "numpy",
+                "array_api_compat.numpy",
+                "cupy",
+                "cupy.array_api",
+                "array_api_strict",
+            }:
+                # XXX: the 'cpu' string isn't portable. Once libraries support the
+                # 2023.12 version of the standard, there is an introspection API
+                # that can be used to check if float64 is supported on the current
+                # device (if it is, the try-except can be avoided here). And the
+                # copy/device keywords to `from_dlpack` can be used for
+                # portable "copy to host" behavior.)
+                sample_score = xp.astype(
+                    xp.asarray(sample_score, device="cpu"), xp.float64
+                )
```

Comment:
This whole branch will go away once #27904 is finalized because the upcast to `float64` will go away.